clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -252,11 +305,38 @@ private void setSaslState(SaslState saslState) {
             this.pendingSaslState = null;
             this.saslState = saslState;
             LOG.debug("Set SASL client state to {}", saslState);
-            if (saslState == SaslState.COMPLETE)
-                transportLayer.removeInterestOps(SelectionKey.OP_WRITE);
+            if (saslState == SaslState.COMPLETE) {
+                setAuthenticationEndAndSessionReauthenticationTimes();
+                /*
+                 * Re-authentication is triggered by a write, so we have to make sure that
+                 * pending write is actually sent.
+                 */
+                transportLayer.addInterestOps(SelectionKey.OP_WRITE);
```

Comment:
This should only be done for re-authentication. Otherwise it will end up with tight poll loop after authentication if there is nothing to be sent. In general, we need to make sure that we don't change any behaviour for the initial authentication.

Reply:
👍 fixed.  I think maybe the original code had kept track of whether we we needed to immediately write something or not, and perhaps it was handled there?  Can't remember, but regardless, it is fixed now -- we remove write interest as before when we are not in a "re-authentication" scenario.